### DISCUSSION TITLE
export useRenderProps

### DISCUSSION THREAD

**Discussion by unional** [adobe/react-spectrum#6443] [Ideas]
I want to build a component using the `slot` mechanism in `react-aria-components`.

When I tried it out, I notice that hooks like `useRenderProps` in `./util` is not exported.

Is it possible to export those hooks and related code?

**Comment by snowystinger** [comment]
For slots, it's probably best to use your own context so that you don't ever conflict with our context.
`useRenderProps` doesn't have anything to do with slots. What are you trying to do there?

It sounds like maybe you're trying to make other RAC API-like components, not building components on top of RAC?

We do want to be a little careful about exporting these, but if you can talk more about your use-case and maybe show some examples, we can take it into consideration. Essentially starting an RFC https://react-spectrum.adobe.com/contribute.html#feature-requests

**Reply by unional** [reply]
I'm trying to build a data grid implementation as the `Table` in `react-aria-components` is built based on `<table>` and doesn't work for nested table (or row expansion to anything) AFAIK.

So I want to use the slots mechanism to do similar things such as drag and drop, selection, etc.

I'm planning to use different context (e.g. `DataGridContext` as compare to `TableContext`) so they won't be any conflict.

I looked at `react-aria-components` to see how the component handles the props, like inherit from the context using `useContextProps`, and handle `className` and `style` resolution using `useRenderProps`. That's when I notice `useRenderProps` and some of the related types such as `StyleRenderProps<T>` are not exported.

There are more code needs to be exported (e.g. `useCollection`), but it really depends on how does the team defines the public API.

**Reply by snowystinger** [reply]
Chatted with the team. We do intend to eventually export most of the Collection work. Hopefully in near future.
We're not sure yet about useRenderProps, however, it's pretty straightforward, you could copy it for now.

**Reply by greg-hoarau** [reply]
Hi ! Regarding useRenderProps, I also want to use it in my project. For now, I'll go ahead and copy it as you suggested 👍 Thanks
